src/Language/PureScript/CoreFn/Traversals.hs

```diff
@@ -13,34 +14,59 @@ everywhereOnValues :: (Bind a -> Bind a) ->
                       (Expr a -> Expr a) ->
                       (Binder a -> Binder a) ->
                       (Bind a -> Bind a, Expr a -> Expr a, Binder a -> Binder a)
-everywhereOnValues f g h = (f', g', h')
+everywhereOnValues f g h =
+  let (f', g', h') = everywhereOnValuesM (return . f) (return . g) (return . h)
+   in (runIdentity . f', runIdentity . g', runIdentity . h')
+
+everywhereOnValuesM :: (Monad m) =>
+                       (Bind a -> m (Bind a)) ->
+                       (Expr a -> m (Expr a)) ->
+                       (Binder a -> m (Binder a)) ->
+                       (Bind a -> m (Bind a), Expr a -> m (Expr a), Binder a -> m (Binder a))
+everywhereOnValuesM f g h = (f', g', h')
   where
-  f' (NonRec a name e) = f (NonRec a name (g' e))
-  f' (Rec es) = f (Rec (map (second g') es))
+  f' (NonRec a name e) = f . NonRec a name =<< g' e
+  f' (Rec es) = f . Rec =<< mapM (\(b, e) -> (b,) <$> g' e) es
 
-  g' (Literal ann e) = g (Literal ann (handleLiteral g' e))
-  g' (Accessor ann prop e) = g (Accessor ann prop (g' e))
-  g' (ObjectUpdate ann obj vs) = g (ObjectUpdate ann (g' obj) (map (fmap g') vs))
-  g' (Abs ann name e) = g (Abs ann name (g' e))
-  g' (App ann v1 v2) = g (App ann (g' v1) (g' v2))
-  g' (Case ann vs alts) = g (Case ann (map g' vs) (map handleCaseAlternative alts))
-  g' (Let ann ds e) = g (Let ann (map f' ds) (g' e))
+  g' (Literal ann e) = g . Literal ann =<< handleLiteral g' e
+  g' (Accessor ann prop e) = g . Accessor ann prop =<< g' e
+  g' (ObjectUpdate ann obj vs) = do
+    obj' <- g' obj
+    vs' <- mapM (mapM g') vs
+    g (ObjectUpdate ann obj' vs')
+  g' (Abs ann name e) = g . Abs ann name =<< g' e
+  g' (App ann v1 v2) = do
+    v1' <- g' v1
+    v2' <- g' v2
+    g (App ann v1' v2')
+  g' (Case ann vs alts) = do
+    vs' <- mapM g' vs
+    alts' <- mapM handleCaseAlternative alts
+    g (Case ann vs' alts')
+  g' (Let ann ds e) = do
+    ds' <- mapM f' ds
+    e' <- g' e
+    g (Let ann ds' e')
   g' e = g e
 
-  h' (LiteralBinder a b) = h (LiteralBinder a (handleLiteral h' b))
-  h' (NamedBinder a name b) = h (NamedBinder a name (h' b))
-  h' (ConstructorBinder a q1 q2 bs) = h (ConstructorBinder a q1 q2 (map h' bs))
+  h' (LiteralBinder a b) = h . LiteralBinder a =<< handleLiteral h' b
+  h' (NamedBinder a name b) = h . NamedBinder a name =<< h' b
+  h' (ConstructorBinder a q1 q2 bs) = h . ConstructorBinder a q1 q2 =<< mapM h' bs
   h' b = h b
 
-  handleCaseAlternative ca =
-    ca { caseAlternativeBinders = map h' (caseAlternativeBinders ca)
-       , caseAlternativeResult = (map (g' *** g') +++ g') (caseAlternativeResult ca)
-       }
+  handleCaseAlternative ca = do
+    caseAlternativeBinders' <- mapM h' (caseAlternativeBinders ca)
+    caseAlternativeResult' <- case caseAlternativeResult ca of
+      Left guardeds -> Left <$> mapM (\(grd, e) -> (,) <$> g' grd <*> g' e) guardeds
```

Comment:
You could use `pairM` from `Language.PureScript.Traversals` here I think, like we do in the `AST` traversals:

https://github.com/purescript/purescript/blob/5aaf615decfe20a84be051e118260ded4743f2b0/src/Language/PureScript/AST/Traversals.hs#L218-L222